DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp

```diff
@@ -5,140 +5,97 @@
+
 // dpc_common.hpp can be found in the dev-utilities include folder.
 // e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
 #include "dpc_common.hpp"
 
-using namespace sycl;
 
-constexpr int kInitNumInputs = 16 * 1024 * 1024;  // Default number of inputs.
-constexpr int kNumOutputs = 64;                   // Number of outputs
-constexpr int kInitSeed = 42;         // Seed for randomizing data inputs
-constexpr int kCacheDepth = 5;        // Depth of the cache.
-constexpr int kNumRuns = 2;           // runs twice to show the impact of cache
-constexpr double kNs = 1000000000.0;  // number of nanoseconds in a second
+constexpr int kInitNumInputs = 16 * 1024 * 1024;  // Default number of inputs
+constexpr int kNumOutputs = 64;           // Number of outputs
+constexpr int kInitSeed = 42;             // Seed for randomizing data inputs
+constexpr int kMaxCacheDepth = MAX_CACHE_DEPTH; // max cache depth to test
+constexpr int kMinCacheDepth = MIN_CACHE_DEPTH; // min cache depth to test
+constexpr double kNs = 1000000000.0;      // number of nanoseconds in a second
 
 // Forward declare the kernel name in the global scope.
 // This FPGA best practice reduces name mangling in the optimization reports.
-template<bool use_cache>
-class Task;
-
-// This kernel function implements two data paths: with and without caching.
-// use_cache specifies which path to take.
-template<bool use_cache>
-void Histogram(sycl::queue &q, buffer<uint32_t>& input_buf,
-               buffer<uint32_t>& output_buf, event& e) {
+template<size_t cache_depth> class HistogramID;
+
+template<size_t k_cache_depth>
+void ComputeHistogram(sycl::queue &q, sycl::buffer<uint32_t>& input_buf,
+                      sycl::buffer<uint32_t>& output_buf, sycl::event& e) {
   // Enqueue  kernel
-  e = q.submit([&](handler& h) {
+  e = q.submit([&](sycl::handler& h) {
     // Get accessors to the SYCL buffers
-    accessor input(input_buf, h, read_only);
-    accessor output(output_buf, h, write_only, no_init);
+    sycl::accessor input(input_buf, h, sycl::read_only);
+    sycl::accessor output(output_buf, h, sycl::write_only, sycl::no_init);
 
-    h.single_task<Task<use_cache>>([=]() [[intel::kernel_args_restrict]] {
+    h.single_task<HistogramID<k_cache_depth>>(
+    [=]() [[intel::kernel_args_restrict]] {
 
       // On-chip memory for Histogram
-      uint32_t local_output[kNumOutputs];
-      uint32_t local_output_with_cache[kNumOutputs];
-
-      // Register-based cache of recently-accessed memory locations
-      uint32_t last_sum[kCacheDepth + 1];
-      uint32_t last_sum_index[kCacheDepth + 1];
-
-      // Initialize Histogram to zero
-      for (uint32_t b = 0; b < kNumOutputs; ++b) {
-        local_output[b] = 0;
-        local_output_with_cache[b] = 0;
-      }
-
+      // A k_cache_depth of 0 is equivalent to a standard array with no cache
+      fpga_tools::OnchipMemoryWithCache<uint32_t, kNumOutputs, k_cache_depth> 
+        histogram(0);
```

Comment:
can you call this `histogram_mem` so it's easier to see later in the code that it's a memory? (I'm stupid and need to be beat over the head with information so it sinks in.)
```suggestion
        histogram_mem(0);
```

Reply:
It's not just a memory though, it's a memory with a cache.

Also, I don't think it's standard practice to add _mem to anything that's a memory.  Most arrays aren't called, for example, q_matrix_mem or steering_vector_mem.  I don't think this would enhance the clarity of the code.